src/btree/bt_compact.c

```diff
@@ -257,12 +256,18 @@ __compact_walk_internal(WT_SESSION_IMPL *session, WT_REF *parent)
     WT_INTL_FOREACH_BEGIN (session, parent->page, ref) {
         if (F_ISSET(ref, WT_REF_FLAG_LEAF)) {
             WT_ERR(__compact_page(session, ref, &skipp));
+            if (session->compact_state == WT_COMPACT_EXITING)
+                break;
             if (!skipp)
                 overall_progress = true;
         }
     }
     WT_INTL_FOREACH_END;
 
+    /* If compaction should quit, we're done. */
+    if (session->compact_state == WT_COMPACT_EXITING)
+        goto err;
```

Comment:
In the loop just above, if `ret` is different than `0`, it would be caught by:
```
WT_ERR(__compact_page(session, ref, &skipp));
```
Therefore, we know it's 0.